### ISSUE TITLE
Snowflake python connector monkeypatches urllib and makes many services unusable.

### ISSUE THREAD

**Issue by potiuk** [apache/airflow#12881]
Curreently wnen you run snowflke provider, it monkeypatches urlllb in a way that is not compatible with other libraries (for example presto SSL with kerberos, google,  amazon, qubole and many others).

This is not critical (as in 2.0 we have provider separation and snowflake code will not even be there until you choose [snowflake] extra or install provider manually,

For now we decided to release but immediately yank the snowflake provider!

Additional links: 

* Issue: https://github.com/snowflakedb/snowflake-connector-python/issues/324

Offending code:

* https://github.com/snowflakedb/snowflake-connector-python/blob/133d6215f7920d304c5f2d466bae38127c1b836d/src/snowflake/connector/network.py#L89-L92

**Comment by potiuk** [comment]
cc: @mik-laj

**Comment by eladkal** [comment]
Raised the issue with my contact at Snowflake

**Comment by lgwacker** [comment]
Seems like a fix is finally on the way https://github.com/snowflakedb/snowflake-connector-python/pull/557

**Comment by ashb** [comment]
I'm still not sure _why_ they are even doing such monkeypatching. Does anyone know why they do this?

**Comment by lgwacker** [comment]
I read all issues related to that there and never found why they did that...
The main issue was open since June. But with community backlash seems like they started working on it. :pray:

**Comment by mik-laj** [comment]
@ashb OCSP support.

**Comment by potiuk** [comment]
Yeah. OCSP support was the root cause, but the solution to monkeypatch a library that others use was, well.. offensive. They could have vendoered it in as we did with some other packages in the past

**Comment by ashb** [comment]
Or submitted the patch upstream :)

**Comment by manugarri** [comment]
fyi, we have been trying to get snowflake to fix this for months.

As an alternative, I made a [lite version of snowflake-connector package](https://github.com/manugarri/snowflake-connector-python-lite) that i use in my airflow environments in case someone wants to check it out

**Comment by potiuk** [comment]
Thanks. I think we are just trying to push snowflake (and we have some heavyweights helping us) to fix it, so let's wait and see :).

**Comment by eladkal** [comment]
**Update:** just heard back from Snowflake. The current estimation is that they will fix it by end of this year.
I'll keep track

**Comment by manugarri** [comment]
great @eladkal . thanks!

**Comment by potiuk** [comment]
Thanks @eladkal !

**Comment by potiuk** [comment]
Yeah, I think NOT switching to a slimmed down version of the lib is a better incentive for Snowflake to fix it :)

**Comment by manugarri** [comment]
agree @potiuk , but i dont work in a heavyweight company so i had to make our data pipelines and reports work.  

I would definitely recommend a full rewrite of the python package, how it handles dependencies is an absolute mess.

**Comment by potiuk** [comment]
Let's see. I hope things will get better with Snowflke's integrations and ecosystem in general. They have a very, very cool product.

**Comment by davido912** [comment]
I would assume this https://github.com/apache/airflow/issues/13255 issue I raised is related (?)

**Comment by potiuk** [comment]
Absolutely @davido912

**Comment by eladkal** [comment]
Snowflake released new version of the connector

https://pypi.org/project/snowflake-connector-python/
```
v2.3.8
Arrow result conversion speed up.
Send all Python Connector exceptions to in-band or out-of-band telemetry.
Vendoring requests and urllib3 to contain OCSP monkey patching to our library only.
Declare dependency on setuptools.
```

**Comment by potiuk** [comment]
Woooho! Switching to it!

**Comment by davido912** [comment]
what a day. let us know if this solved anything! :D

**Comment by potiuk** [comment]
Thanks  @eladkal for keeping an eye :+1:

**Comment by potiuk** [comment]
@eladkal - absolutelty https://github.com/apache/airflow/pull/13654 - I unmarked the test in presto that was marked as "expect to fail" because of this problem, this will tell us whether the problem is fixed.

**Comment by manugarri** [comment]
I dont think this PR fixes the strict dependencies issue though https://github.com/snowflakedb/snowflake-connector-python/blob/master/setup.py#L194 , that is usually what breaks builds.

**Comment by potiuk** [comment]
> I dont think this PR fixes the strict dependencies issue though https://github.com/snowflakedb/snowflake-connector-python/blob/master/setup.py#L194 , that is usually what breaks builds.

Not really. You do not have to install the snowflake provider (Airflow in 2.0 is now split into separate core and providers). Even in 1.10 snowflake was an 'extra' that you did not have to install. But in the meantime we impleemented much better way of handling the requirement upgrade, so now we have less requirements in airflow's install_requires and those that are problematic are moved out of airflow 'install_requires' into test Dockerfiles when we could. 

Thanks to that 2.0.1 and 1.10.15 should be much gentler w/regards to requirements:

See here: https://github.com/apache/airflow/blob/bfb7cb343d0adfdca92a77a273954d9f2dbcfdc9/Dockerfile.ci#L345 for those problematic requirements. Airflow on it's own does not depend any more on them, so as long as you are not installing any "problematic' providers you will be able to upgrade.

**Comment by manugarri** [comment]
@potiuk by strict dependencies issue i meant snowflake-python-connector dependencies, not Airflow's

**Comment by potiuk** [comment]
> @potiuk by strict dependencies issue i meant snowflake-python-connector dependencies, not Airflow's

I see. They got improved but indeed could be more open. For me it is a bit less of an issue so far.

**Comment by manugarri** [comment]
@potiuk what that means is, if we install snowflake as an extra dependency in airflow, there might (probably?) be conflicts with other dependencies (boto3 being the usual suspect).

**Comment by potiuk** [comment]
> @potiuk what that means is, if we install snowflake as an extra dependency in airflow, there might (probably?) be conflicts with other dependencies (boto3 being the usual suspect).

It is not a problem for airflow - but if you have problem with that - I encourage you to open an issue for that, they seem to be responsive to community complaints

**Comment by manugarri** [comment]
@potiuk that is true, however it means snowflake operators cant be reliably used right? 

And regarding opening issues and responsiveness, well, lets say I opened a [related issue almost a year ago](https://github.com/snowflakedb/snowflake-connector-python/issues/284). In fact i released  [snowflake-connector-python-lite](https://pypi.org/project/snowflake-connector-python-lite/), a lite version of snowflake package just so my company's CI/CD builds would stop failing.

**Comment by potiuk** [comment]
> @potiuk that is true, however it means snowflake operators cant be reliably used right?

Why? I think the dependencies that are left are pretty ok and we are also upgrading azure provider so that the last conflicts are removed. From what I see boto dependency in snowflake connector is:

```
boto3>=1.4.4,<2.0.0
```


So nothing extraodirinary -  taking into account that latest released version is 1.16.56. 
Do you KNOW that you have some other conflicts, or do you just suspect it ?

From the latest set of constraints generated for Airflow (I just merged the snowflake change and regenerated constraints) - we have those constraints 'blessed" for airflow (3.6 example):

https://github.com/apache/airflow/blob/constraints-master/constraints-3.6.txt

* boto3==1.15.18 (not latest but this is because amazon provider limits it (`boto3>=1.15.0,<1.16.0`) - if amazon provider is 
                             upgraded, it will go up
* snowflake-connector-python==2.3.8
* requests==2.25.1 (latest)
* urllib3==1.25.11 (not latest but it is not limited by python-snowflake-connector) - actually one of my fixes to the connector
                             which I contributed to snowflake was getting rid of the urllib3 dependency altogether as they did not need it.
                             The < 1.26 comes from boto-core actually, so if amazon provider goes up with botocore, this one will go up 
                             as well.

Do you think your problems are not solved by this?
